Refactoring around the goto-cc entry point.

thomasspriggs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

As I was working on https://github.com/diffblue/cbmc/pull/5440, I found the entry point code less than straight forward to follow. I carried out this refactorings locally in order to make it easier for me to understand. I am raising this separate PR containing the refactoring work only as it was not required in order to implement the feature in the original PR, but should still improve maintainability.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] N/A - No user visible behaviour changes. ~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] N/A - None claimed ~My commit message includes data points confirming performance improvements (if claimed).~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Make `goto_cc_modet::help` non-virtual

Making this member function non-virtual makes it easier to follow the
control flow because it is clear it not overridden in any sub class.
This is should be easy enough to add back should a need to override this
arise.

thomasspriggs committed:
Move `prefix_in_list` to the single location where it is used

This makes the `goto_cc_cmdlinet` class easier to follow because it can
be read without seeing this special case for `armcc`.

thomasspriggs committed:
Refactor `prefix_in_list` to return its result using `optionalt`

This avoids using an output parameter, which makes it more straight
forward to read the data flow.

thomasspriggs committed:
Move `goto_cc_cmdlinet::have_infile_arg` to `.cpp`

Putting implementations in `.cpp` files is the best-practise because
it reduces compile times.

thomasspriggs committed:
Refactor `goto_cc_cmdlinet::have_infile_arg` using `std::any_of`

This avoids the need to declare and mutate iterators. Its states what it
is doing by means of the name of the algorithm - `std::any_of`, which
saves on reading a loop to work it out.

hannes-steffenhagen-diffblue commented on src/goto-cc/armcc_cmdline.cpp:
```diff
@@ -265,6 +265,21 @@ static const char *options_with_arg[]=
 };
 
+static bool
+prefix_in_list(const char *option, const char **list, std::string &prefix)
```
I think we could simplify a lot of this if we stored the args in the mode as `std::vector<std::string>` instead of `char **`. Then we could also just use regular `std::find` or something like that instead of this.

thomasspriggs replied:
That sounds like a good follow up refactoring.

thomasspriggs replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-cc/armcc_cmdline.cpp:
```diff
@@ -14,6 +14,8 @@ Author: Daniel Kroening
 #include <cstring>
 #include <iostream>
 
+#include "optional.h"
```
should be `<util/optional.h>`

thomasspriggs replied:
Fixed.

hannes-steffenhagen-diffblue commented on src/goto-cc/armcc_cmdline.cpp:
```diff
@@ -265,19 +267,18 @@ static const char *options_with_arg[]=
 
-static bool
-prefix_in_list(const char *option, const char **list, std::string &prefix)
+static optionalt<std::string>
```
Like I said above, I'd prefer not touching this like this for now and instead changing how we store command line arguments, then we don't have to invent our own finds at all.

thomasspriggs replied:
Yes, I think that using a standard container would be a great follow-up refactoring. However I would argue that removing the need for an output parameter is already a step in the right direction, which makes the usages of this function more readable.

thomasspriggs replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-cc/goto_cc_cmdline.cpp:
```diff
@@ -117,3 +117,13 @@ void goto_cc_cmdlinet::add_infile_arg(const std::string &arg)
   }
 }
+
+bool goto_cc_cmdlinet::have_infile_arg() const
```
I think we should probably just make sure to put definitions in cpp files in general.

thomasspriggs replied:
:+1:

hannes-steffenhagen-diffblue commented on src/util/bv_encoding.h:
```diff
@@ -0,0 +1,23 @@
+
+Author: Diffblue Ltd
+
+\*******************************************************************/
```
I think I mentioned this before but we want to move away from this style of module header... we might have to adjust our linter for that though.

thomasspriggs replied:
I have removed this commit from this PR.

hannes-steffenhagen-diffblue commented on src/util/bv_encoding.h:
```diff
@@ -0,0 +1,23 @@
+
+#include <cstddef>
+
+struct bv_encodingt
```
Moves it into a different namespace though

thomasspriggs replied:
Ok. I'll leave it where it was then.

hannes-steffenhagen-diffblue commented on src/util/bv_encoding.h:
```diff
@@ -11,7 +11,7 @@ Author: Diffblue Ltd
 #include <cstddef>
 
-struct bv_encodingt
+struct bv_encodingt final
```
I'm not sure that's necessary, I find `final` to be a bit noisy. To me `final` actually suggests the opposite of what you're saying here: This *is* part of a class hierarchy, it's just supposed to be at the bottom of it.

thomasspriggs replied:
You know that isn't part of a hierarchy from reading this line of code, because no base classes are specified. Having it specified as `final` also means that you are always safe to pass / return / assign instances of this class by value, without risk of slicing occurring. I do take your point about noise however. I think I may drop this commit as I don't really want to include anything in this PR, where the pros and cons could be argued either way. Suffice to say that my current habit is to make new classes/structs `final` by default, in the same way as I would make variables and member functions `const` by default.

hannes-steffenhagen-diffblue commented on src/util/bv_encoding.cpp:
```diff
@@ -0,0 +1,31 @@
+  const std::string &argument,
+  const std::size_t pointer_width)
+{
+  const unsigned int object_bits = unsafe_string2unsigned(argument);
```
Use `string2optional` here so you can handle a parse error with a proper error message. The `unsafe_` functions are for when you know the parse will succeed.

thomasspriggs replied:
Yep. A non-integer argument will cause an invariant violation, both before and after this PR. This is expected as the PR is intended to be refactoring only, not bug fixing. I agree that we should improve the error handling here.

thomasspriggs committed:
Make `goto_cc_modet::main` non-virtual

Making this member function non-virtual makes it easier to follow the
control flow because it is clear it not overridden in any sub class.
This is should be easy enough to add back should a need to override this
arise.

thomasspriggs committed:
Refactor `prefix_in_list` to take vector of strings

Because this gives us compatability with modern C++ features such as
ranged for loops and the algorithms in the standard template library.
Because using a standard container rather than a null terminated array
improves memory safety.

Clang format is off for the lists of options in order to keep the
existing manual formatting. This existing formatting is consistent with
the other lists of options.

thomasspriggs committed:
Add unit test of `prefix_in_list`

To demonstrate that the refactored version works as expected.

chrisr-diffblue commented on unit/goto-cc/armcc_cmdline.cpp:
```diff
@@ -0,0 +1,32 @@
+// Copyright 2020 Diffblue Limited. All Rights Reserved.
```
I think the licence header is wrong - thats the internal Diffblue header - CBMC contributions require a different header.

chrisr-diffblue reviewed (commented):
All looks good apart from the copyright header - happy to approve once that is fixed up.
